Regression: .global initialization inside method hoisted to preInitProc without temporary variable declarations (invalid C codegen)

### Nim Version

```console
$ nim -v
Nim Compiler Version 2.2.6 [Linux: amd64]
Compiled at 2025-10-31
Copyright (c) 2006-2025 by Andreas Rumpf

git hash: ab00c56904e3126ad826bb520d243513a139436a
active boot switches: -d:release
```

This issue does not occur in 2.2.4, and appears starting from 2.2.6 (regression).

### Description

When a method contains a let binding with a {.global.} pragma and a nested call like f(f("")), Nim 2.2.6 generates invalid C code.

**Minimal Reproducible Example**
```nim
proc f(str: string): string = discard
method m(self: RootObj) {.base.} =
  let v {.global, used.}: string = f(f(""))
```

This only happens in very specific conditions:

- method (not proc)
- a let binding with {.global.}
- nested function call produces temporaries (e.g. f(f("")))
- the declaration and initialization happen in one statement

Any of the following avoids the bug:

- lowering the nesting depth f(f("")) → f("")
- changing method to proc
- removing {.global.}
- separating declaration and initialization into two statements

**What seems to be happening**
In Nim 2.2.6, the initialization of the {.global.} variable is hoisted from inside the method into the preInitProc section of NimMainModule.

However, the temporary variable (colontmpD_) used in the nested call f(f("")) is not hoisted with it, resulting in the generated C code referencing an undeclared identifier.

### Current Output

```text
/home/panno/.cache/nim/test_d/@mtest.nim.c: In function ‘NimMainModule’:
/home/panno/.cache/nim/test_d/@mtest.nim.c:178:9: error: ‘colontmpD_’ undeclared (first use in this function)
  178 |         colontmpD_ = f__test_u1(TM__ipcYmBC9bj9a1BW35ABoB1Kw_3);
      |         ^~~~~~~~~~
```

### Expected Output

```text
The program should compile successfully as it did in Nim 2.2.4
```

### Known Workarounds

**lowering the nesting depth f(f("")) → f("")**
```nim
proc f(str: string): string = discard
method m(self: RootObj) {.base.} =
  let v {.global, used.}: string = f("")
```

**changing method to proc**
```nim
proc f(str: string): string = discard
proc m() =
  let v {.global, used.}: string = f(f(""))
```

**removing {.global.}**
```nim
proc f(str: string): string = discard
method m(self: RootObj) {.base.} =
  let v {.used.}: string = f(f(""))
```

**separating declaration and initialization into two statements**
```nim
proc f(str: string): string = discard
method m(self: RootObj) {.base.} =
  var v {.global.}: string
  once: v = f(f(""))
```

### Additional Information

_No response_

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -2449,7 +2449,10 @@ proc handleProcGlobals(m: BModule) =
 
     # fixes recursive calls #24997
     swap stmts, m.preInitProc.s(cpsStmts)
-    genStmts(m.preInitProc, procGlobals[i])
+    var transformedN = procGlobals[i]
+    if sfInjectDestructors in m.module.flags:
+      transformedN = injectDestructorCalls(m.g.graph, m.idgen, m.module, transformedN)
+    genStmts(m.preInitProc, transformedN)
     swap stmts, m.preInitProc.s(cpsStmts)
 
     handleProcGlobals(m)

--- a/compiler/injectdestructors.nim
+++ b/compiler/injectdestructors.nim
@@ -968,10 +968,10 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSing
                       {sfPure, sfGlobal} <= v.sym.flags and
                       isInProc
 
-              let value = moveOrCopy(v, ri, c, s, if v.kind == nkSym: {IsDecl} else: {})
               if isGlobalPragma:
-                c.graph.procGlobals.add value
+                c.graph.procGlobals.add n
               else:
+                let value = moveOrCopy(v, ri, c, s, if v.kind == nkSym: {IsDecl} else: {})
                 result.add value
             elif ri.kind == nkEmpty and c.inLoop > 0:
               let skipInit = v.kind == nkDotExpr and # Closure var

--- a/tests/global/tglobal3.nim
+++ b/tests/global/tglobal3.nim
@@ -55,3 +55,10 @@ block: # bug #24997
       doAssert not isNil(u(typeof(B.j)))
     R()
   discard u(B)
+
+proc f2(str: string): string = str
+proc m2()  =
+  let v {.global, used.}: string = f2(f2("123"))
+  assert v == "123"
+
+m2()

